tools/fork-network/src/cli.rs

```diff
@@ -522,23 +597,156 @@ impl ForkNetworkCommand {
         let new_state_roots = update_state.into_iter().map(|u| u.state_root()).collect::<Vec<_>>();
         tracing::info!("Creating a new genesis");
         backup_genesis_file(home_dir, &near_config)?;
+
+        // 4. Create new genesis with updated state roots and validators.
+        let mut original_genesis_config = near_config.genesis.config.clone();
+        let genesis_file = near_config.config.genesis_file.clone();
+        original_genesis_config.chain_id = chain_id.clone();
+        original_genesis_config.genesis_time = genesis_time;
+        original_genesis_config.protocol_version = protocol_version;
+        original_genesis_config.genesis_height = flat_head.height + 1;
+        original_genesis_config.epoch_length = epoch_length;
         self.make_and_write_genesis(
+            home_dir,
+            original_genesis_config,
+            genesis_file,
+            epoch_config,
+            new_state_roots,
+            new_validator_accounts,
+        )
+    }
+
+    /// Reads configuration patches for sharded benchmark.
+    /// For now reads only epoch config and number of accounts per shard to
+    /// benchmark.
+    fn read_patches(patches_path: &Path) -> anyhow::Result<(EpochConfig, u64)> {
+        let epoch_config_path = patches_path.join("epoch_configs/template.json");
+        let epoch_config: EpochConfig =
+            serde_json::from_str(&std::fs::read_to_string(epoch_config_path)?)?;
+        let params_path = patches_path.join("params.json");
+        let params: HashMap<String, serde_json::Value> =
+            serde_json::from_str(&std::fs::read_to_string(params_path)?)?;
+        let num_accounts = params["num_accounts"].as_u64().unwrap();
+        Ok((epoch_config, num_accounts))
+    }
+
+    /// Sets genesis block to be able to write accounts to the state.
+    fn set_genesis_block(
+        epoch_manager: &dyn EpochManagerAdapter,
+        runtime: &dyn RuntimeAdapter,
+        chain_genesis: &ChainGenesis,
+        state_roots: Vec<StateRoot>,
+    ) -> anyhow::Result<()> {
+        let (genesis_block, _) =
+            Chain::make_genesis_block(epoch_manager, runtime, &chain_genesis, state_roots)?;
+        let flat_storage_manager = runtime.get_flat_storage_manager();
+        let mut store_update = runtime.store().store_update();
+        let shard_uids =
+            epoch_manager.get_shard_layout(&EpochId::default())?.shard_uids().collect::<Vec<_>>();
+        for shard_uid in shard_uids {
+            flat_storage_manager.set_flat_storage_for_genesis(
+                &mut store_update.flat_store_update(),
+                shard_uid,
+                genesis_block.hash(),
+                genesis_block.header().height(),
+            );
+        }
+        store_update.commit()?;
+        Ok(())
+    }
+
+    fn set_validators(
```

Comment:
Nit: It seems there's some overlap between what this function does and what `set_validators_from_dump`.

I don't have a concrete suggestion , aside from trying to extract some common stuff in smaller functions. Maybe not worth it

Reply:
I don't think overlap is that big. While we call same functions, we use different arguments, and the extraction has already been done in `override_epoch_configs`, `ShardUpdateState`, `make_and_write_genesis`.